### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by aWiedebusch**
Austin Wiedebusch
40340720

I have it working at the time of the pull request, but I think I can make improvements on ex03 based on where I concatenate and some elixir practices that make it more readable.  Will update later if I get around to it.

### CODE REVIEW

**Comment by aWiedebusch** [comment]
Made it much easier to read and much more elixir(y) 

Note: I had to raise the number of processes to 11_000_000 to get it to consistently pass the test.

File: ex01.exs [hunk 1]
```diff
@@ -28,6 +28,24 @@ defmodule Ex01 do
   """
   
   def counter(value \\ 0) do
+    receive do
+      {:next, from} -> 
+        send from, { :next_is, value }
+        value = value + 1
```
Comment (pragdave) [inline ex01.exs:34]: ## -2

This is going to seem very picky, but

~~~ elixir
value = value + 1
counter(value)
~~~

expresses (to me, at least) a different meaning to

~~~ elixir
counter(value + 1)
~~~

The first form goes to the trouble of assigning to a variable, which suggests that this value will be needed later. But, later is just the next line, and a simple increment doesn't complicate this line.

So the second form expresses what you are doing more clearly—call myself with an incremented value.

You could also express this using

~~~ elixir
value + 1 |> counter()
~~~

File: ex02.exs [hunk 2]
```diff
@@ -1,6 +1,25 @@
 
 defmodule Ex02 do
 
+  def new_counter(value) do
+    {:ok, count} = Agent.start(fn -> value - 1 end)
+    count
+  end
+
+  def next_value(count) do
+    Agent.update(count, &(&1 + 1))
```
Comment (pragdave) [inline ex02.exs:10]: ## -4

This is actually a bug—there's a race condition. Two different processes could call this function at the same time. The first does the update, and then the scheduler suspends it and runs the second. It also does the update and then gets the updated value. The first process then gets scheduled and gets the value, but by this time it has been incremented twice.

This is why there's `Agent.get_and_update`

File: ex03.exs [hunk 3]
```diff
@@ -60,9 +60,18 @@ defmodule Ex03 do
   """
 
   def pmap(collection, process_count, function) do
-    « your code here »
-  end
+    size = Enum.count(collection) / process_count
```
Comment (pragdave) [inline ex03.exs:63]: No deduction, but you can also use `div` for integer division.

**Comment by pragdave** [comment]
Dialing the tests up to 11.... :)

**Pull request closed without merging**